File: server-node/src/handlers/license.ts
/**
 * License activation endpoint.
 *
 *   POST /api/v1/license/activate
 *     { key, fingerprint, machine_label? }
 *   →
 *     200 { token_json: "<JSON-encoded LicenseToken>" }
 *     404 LICENSE_NOT_FOUND
 *     409 DEVICE_BOUND
 *     410 LICENSE_REVOKED
 *
 * No unbind path exists — licenses are bound to a single device for life.
 * The only re-bind allowed is when the same fingerprint comes back (e.g.
 * reinstall) — that path returns a freshly signed token.
 */

import { Hono } from "hono";
import { z } from "zod";
import type { AppEnv } from "../app.js";
import { signToken } from "../license/sign.js";

const TRIAL_DAYS = 5;

const ActivateSchema = z.object({
  key: z.string().min(1),
  fingerprint: z.string().min(16),
  machine_label: z.string().optional().nullable(),
});

const StartTrialSchema = z.object({
  fingerprint: z.string().min(16),
  machine_label: z.string().optional().nullable(),
});

export const licenseRouter = new Hono<AppEnv>();

licenseRouter.post("/license/activate", async (c) => {
  const parsed = ActivateSchema.safeParse(await c.req.json().catch(() => ({})));
  if (!parsed.success) {
    return c.text("INVALID_REQUEST", 400);
  }

  const key = parsed.data.key.trim().toUpperCase();
  const fingerprint = parsed.data.fingerprint.toLowerCase();
  const label = parsed.data.machine_label ?? "";

  const { db, signingKey } = c.var.app;

  type Row = {
    plan: string;
    bound_fingerprint: string | null;
    revoked: number;
  };
  const row = db
    .prepare(
      `SELECT plan, bound_fingerprint,
              CASE WHEN COALESCE(note, '') LIKE 'REVOKED%' THEN 1 ELSE 0 END AS revoked
         FROM licenses WHERE key = ?`,
    )
    .get(key) as Row | undefined;

  if (!row) return c.text("LICENSE_NOT_FOUND", 404);
  if (row.revoked) return c.text("LICENSE_REVOKED", 410);

  // Binding check
  if (row.bound_fingerprint && row.bound_fingerprint.toLowerCase() !== fingerprint) {
    return c.text("DEVICE_BOUND", 409);
  }

  if (!row.bound_fingerprint) {
    db.prepare(
      `UPDATE licenses
          SET bound_fingerprint = ?, bound_at = datetime('now'), machine_label = ?
        WHERE key = ?`,
    ).run(fingerprint, label, key);
  }

  const issuedAt = new Date();
  const plan = row.plan === "trial" ? "trial" : "lifetime";
  const expiresAt =
    plan === "trial" ? new Date(issuedAt.getTime() + 5 * 86400 * 1000) : null;

  const tokenJson = await signToken(signingKey.privateKey, {
    key,
    plan,
    fingerprint,
    issuedAt,
    expiresAt,
  });

  return c.json({ token_json: tokenJson });
});

/**
 * Start (or re-issue) a 5-day Pro trial for the given hardware fingerprint.
 *
 * The server is the **only** place trial eligibility is tracked. Each
 * fingerprint can start exactly one trial; deleting local files on the
 * client does not reset it. If a trial is already on file:
 *   - still inside its 5-day window → re-sign a token with the SAME
 *     expires_at (idempotent for re-installs)
 *   - already expired → 409 TRIAL_ALREADY_USED
 */
licenseRouter.post("/license/start_trial", async (c) => {
  const parsed = StartTrialSchema.safeParse(await c.req.json().catch(() => ({})));
  if (!parsed.success) return c.text("INVALID_REQUEST", 400);

  const fingerprint = parsed.data.fingerprint.toLowerCase().trim();
  const machineLabel = parsed.data.machine_label ?? "";
  const ip =
    c.req.header("x-forwarded-for")?.split(",")[0]?.trim() ??
    c.req.header("x-real-ip") ??
    "?";
  const ua = c.req.header("user-agent") ?? "";

  const { db, signingKey } = c.var.app;

  type Row = { started_at: string; expires_at: string };
  const existing = db
    .prepare(`SELECT started_at, expires_at FROM trials WHERE fingerprint = ?`)
    .get(fingerprint) as Row | undefined;

  let issuedAt: Date;
  let expiresAt: Date;

  if (existing) {
    const exp = new Date(existing.expires_at);
    if (Number.isNaN(exp.getTime()) || exp < new Date()) {
      return c.text("TRIAL_ALREADY_USED", 409);
    }
    // Same fingerprint, trial still active — re-issue token with the same
    // expires_at so reinstalls don't get "free" extra days.
    issuedAt = new Date();
    expiresAt = exp;
  } else {
    issuedAt = new Date();
    expiresAt = new Date(issuedAt.getTime() + TRIAL_DAYS * 86400 * 1000);
    db.prepare(
      `INSERT INTO trials (fingerprint, started_at, expires_at, machine_label, ip, user_agent)
       VALUES (?, ?, ?, ?, ?, ?)`,
    ).run(
      fingerprint,
      issuedAt.toISOString(),
      expiresAt.toISOString(),
      machineLabel,
      ip,
      ua,
    );
  }

  // The token's `key` field is required by the schema; for trials we use a
  // synthetic prefix so admin UIs can tell them apart from purchased keys.
  const tokenJson = await signToken(signingKey.privateKey, {
    key: `TRIAL-${fingerprint.slice(0, 16).toUpperCase()}`,
    plan: "trial",
    fingerprint,
    issuedAt,
    expiresAt,
  });
  return c.json({ token_json: tokenJson });
});

File: server/src/handlers/license.rs
//! License activation endpoint.
//!
//! POST /api/v1/license/activate
//!   { "key": "DM-XXXX-XXXX-XXXX-XXXX-XXXX", "fingerprint": "<hex>", "machine_label"?: "..." }
//!   →
//!   200 { "token_json": "<JSON-encoded LicenseToken>" }
//!   404 LICENSE_NOT_FOUND
//!   409 DEVICE_BOUND  (already bound to a different machine)
//!   410 LICENSE_REVOKED
//!
//! No unbinding logic exists — the spec is explicit that licenses are bound
//! to a single device for life. The only re-bind allowed is when the same
//! fingerprint comes back (e.g. after a reinstall) — that path returns the
//! same signed token.

use axum::{extract::State, http::StatusCode, Json};
use chrono::Utc;
use serde::{Deserialize, Serialize};

use crate::license::{sign, TokenPlan};
use crate::AppState;

#[derive(Debug, Deserialize)]
pub struct ActivateRequest {
    pub key: String,
    pub fingerprint: String,
    #[serde(default)]
    pub machine_label: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ActivateResponse {
    pub token_json: String,
}

pub async fn activate(
    State(state): State<AppState>,
    Json(req): Json<ActivateRequest>,
) -> Result<Json<ActivateResponse>, (StatusCode, String)> {
    if req.key.trim().is_empty() {
        return Err((StatusCode::BAD_REQUEST, "key required".into()));
    }
    if req.fingerprint.len() < 16 {
        return Err((StatusCode::BAD_REQUEST, "fingerprint required".into()));
    }

    let key = req.key.trim().to_uppercase();
    let fingerprint = req.fingerprint.to_lowercase();
    let label = req.machine_label.unwrap_or_default();

    let conn = state.db.lock().map_err(internal)?;

    // 1. Find license
    #[derive(Debug)]
    struct Row {
        plan: String,
        bound_fp: Option<String>,
        revoked: bool,
    }
    let row: Option<Row> = conn
        .query_row(
            "SELECT plan, bound_fingerprint, COALESCE(note, '') LIKE 'REVOKED%' \
             FROM licenses WHERE key = ?1",
            [&key],
            |r| {
                Ok(Row {
                    plan: r.get(0)?,
                    bound_fp: r.get(1)?,
                    revoked: r.get(2)?,
                })
            },
        )
        .ok();

    let row = match row {
        Some(r) => r,
        None => return Err((StatusCode::NOT_FOUND, "LICENSE_NOT_FOUND".into())),
    };

    if row.revoked {
        return Err((StatusCode::GONE, "LICENSE_REVOKED".into()));
    }

    // 2. Binding check
    if let Some(bound) = &row.bound_fp {
        if !bound.eq_ignore_ascii_case(&fingerprint) {
            return Err((StatusCode::CONFLICT, "DEVICE_BOUND".into()));
        }
        // Same fingerprint re-activating — sign a fresh token and return it.
    } else {
        // First activation — bind it now.
        let now = Utc::now().to_rfc3339();
        conn.execute(
            "UPDATE licenses SET bound_fingerprint = ?1, bound_at = ?2, machine_label = ?3 \
             WHERE key = ?4",
            rusqlite::params![fingerprint, now, label, key],
        )
        .map_err(internal)?;
    }

    // 3. Sign and return.
    let plan = match row.plan.as_str() {
        "lifetime" => TokenPlan::Lifetime,
        "trial" => TokenPlan::Trial,
        other => {
            tracing::warn!("unknown license plan {} for key {}", other, key);
            TokenPlan::Lifetime
        }
    };
    let issued = Utc::now();
    let expires = if matches!(plan, TokenPlan::Trial) {
        Some(issued + chrono::Duration::days(5))
    } else {
        None
    };
    let token = sign::sign_token(&state.signing_key, &key, plan, &fingerprint, issued, expires);
    let token_json = serde_json::to_string(&token).map_err(internal)?;

    Ok(Json(ActivateResponse { token_json }))
}

fn internal<E: std::fmt::Display>(e: E) -> (StatusCode, String) {
    tracing::error!("activate: {}", e);
    (StatusCode::INTERNAL_SERVER_ERROR, "internal error".into())
}
